tests/kafkatest/tests/streams/streams_upgrade_test.py

```diff
@@ -533,12 +521,34 @@ def do_rolling_bounce(self, processor, counter, current_generation):
                                            err_msg="Could not detect 'Triggering new rebalance' at upgrading node " + str(node.account))
 
                     # version probing should trigger second rebalance
-                    current_generation = current_generation + 1
-
-                    for p in self.processors:
-                        monitors[p].wait_until("Successfully joined group with generation " + str(current_generation),
-                                               timeout_sec=60,
-                                               err_msg="Never saw output 'Successfully joined group with generation " + str(current_generation) + "' on" + str(p.node.account))
+                    # now we check that after consecutive rebalances we have synchronized generation
+                    generation_synchronized = False
+                    retries = 0
+
+                    while retries < 10:
+                        processor_found = list(node.account.ssh_capture("grep \"Successfully joined group with generation\" %s | awk \'{for(i=1;i<=NF;i++) {if ($i == \"generation\") beginning=i+1; if($i== \"(org.apache.kafka.clients.consumer.internals.AbstractCoordinator)\") ending=i }; for (j=beginning;j<ending;j++) printf $j; printf \"\\n\"}\'" % processor.LOG_FILE, allow_fail=True))
+                        first_other_processor_found = list(first_other_node.account.ssh_capture("grep \"Successfully joined group with generation\" %s | awk \'{for(i=1;i<=NF;i++) {if ($i == \"generation\") beginning=i+1; if($i== \"(org.apache.kafka.clients.consumer.internals.AbstractCoordinator)\") ending=i }; for (j=beginning;j<ending;j++) printf $j; printf \"\\n\"}\'" % first_other_processor.LOG_FILE, allow_fail=True))
+                        second_other_processor_found = list(second_other_node.account.ssh_capture("grep \"Successfully joined group with generation\" %s | awk \'{for(i=1;i<=NF;i++) {if ($i == \"generation\") beginning=i+1; if($i== \"(org.apache.kafka.clients.consumer.internals.AbstractCoordinator)\") ending=i }; for (j=beginning;j<ending;j++) printf $j; printf \"\\n\"}\'" % second_other_processor.LOG_FILE, allow_fail=True))
```

Comment:
I think it might be worth de-magicifying this shell command a little.

Maybe some of the following?
* declare a variable or method for the string and reference it instead of having three copies (the logic is not trivial, and a typo in one would be both bad and hard to detect) (also, it makes it clear that the command is the same for all three nodes)
* add a brief comment explaining what it's doing (we don't need an explanation of exactly what is happening, since we can read the `awk` manual, but it would help to just have a high level description) (I guess if you declare a variable or method, the name of it could also serve this purpose)
* use triple-quotes at the outermost level so you don't have to escape so much in the string